include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td

```diff
@@ -350,4 +352,21 @@ def RegResetWithOneReset : Pat<
   (DropWrite $reg, (NodeOp $resetVal, $name, $nameKind, $annotations, $inner_sym)),
   [(OneConstantOp $reset)]>;
 
+// refresolve(refsend(x)) -> x
+def RefResolveOfSend : Pat<
+  (RefResolveOp (RefSendOp $orig)),
+  (replaceWithValue $orig) >;
+
+// Sink ref.sub through resolve.
+// refresolve(refsub(x, idx)) -> sub{index,field}(refresolve(x))
+def RefResolveOfSub : Pat<
+  (RefResolveOp (RefSubOp $ref, $idx)),
+  (SubFieldOrIndex (RefResolveOp $ref), $idx) >;
```

Comment:
This rewrite seems slightly counterintuitive as it moves hardware out of ref type space and into real hardware space. I think the intention here is that this will then open up opportunities for the real hardware folders, canonicalizers, and CSE to kick in?

I think this is okay, but it does introduce some "cost". Granted this "cost" is likely never realized in hardware given that ref types aren't used in the actual hardware.

Was this motivated by a concrete use case?

Same questions about the `sub(send(..))` below.

Reply:
Great point and thanks for bringing this up!

Mostly ref.sub is a pain to work with and barely supported in CIRCT presently.  And together these make it possible to optimize a `ref.resolve(ref.sub(ref.send(_),idx))`.

So yes, idea is to minimize what's in "ref space" in at least the easy cases-- none of these help if you're doing the `ref.sub` and passing it further along (we'll see how popular it'll be to forward up a portion of an ref-to-aggregate your children passed up to you).

I'm not sure it's necessarily cheaper to do the indexing in the ref space, although on the resolve end that does seem kinda non-obviously-good, it causes the XMR to consume "more" than it would otherwise.  Maybe we want to go the other way here?  If a ref.resolve is only used in part, narrow the reference instead.

I think hoisting the ref.sub out of ref.send side is "strictly" better re:cost, as it means we are creating a reference to less "stuff" and then we can optimize away the rest (instead having to ensure the entire "referent" is around for access).

No concrete use case, just my suite of test inputs for references and being tired of my test cases having trivial local "XMR's" in the final result.  This does help get simple examples involving indexing into references working w/aggregate preservation before we support ref.sub in LowerTypes and before LowerXMR supports `ref.sub`s along the reference dataflow paths.  I think an interprocedural version that hoists the `ref.sub` to eagerly narrow references would be useful, with goal of eventually being able to narrow references to things that don't need the full width.

FWIW, this was dipping toes into the water re:adding a simple version of something stronger that'll need to be added in near future: the ability to gather/move `ref.sub` to one of the `ref.send`/`ref.resolve` endpoints in order for LowerXMR to work.  LowerXMR might be the one gathering them of course, but the point is all sub-{index/field}-ness will need to occur at one of the endpoints since `ref.sub` surely can't manifest in the output Verilog.  However, since LowerXMR runs late in the pipeline I'm thinking it might be good to eagerly handle this to avoid overly pessimizing things we have references to, perhaps as part of rewriting away the necessarily useless input references (for better behavior re:Dedup and general sanity, and as convenient location to reject invalid input reference uses where such a rewrite isn't possible).

Anyway, the above doesn't really necessarily motivate adding it as-is to CIRCT presently, just sharing where things are/may go.

WDYT about favoring hoisting "sub-*"s up towards the ref.send (into ref.resolve, out of ref.send)?

Anyway, I think these aren't careful enough w/handling other users and so need to be fixed/reworked anyway.

Thanks for the review!

Reply:
Trying to apply consistent logic: bubbling toward the ref send makes sense under a cost metric of decreasing connectivity. Generally it seems like less dataflow is always better. (This is obviously true for normal ports ignoring interface contracts between modules.)